File: Go/exercicio4.go
package main

import "fmt"

type Animal interface {
    Som() string
}

type Cachorro struct{}
func (Cachorro) Som() string {
    return "Au Au"
}

type Gato struct{}
func (Gato) Som() string {
    return "Miau"
}

func main() {
    var animais []Animal = []Animal{Cachorro{}, Gato{}}
    for _, animal := range animais {
        fmt.Println(animal.Som())
    }
}

File: Python/exercicio4.py
class Animal:
    def som(self):
        pass

class Cachorro(Animal):
    def som(self):
        return "Au Au"

class Gato(Animal):
    def som(self):
        return "Miau"

animais = [Cachorro(), Gato()]
for animal in animais:
    print(animal.som())